### PR TITLE
Test filtered responses and implement Cors Check Fetch step

### PR DESCRIPTION
**Pull request by nikkisquared**
I've been writing tests for creating filtered responses. So far I have three of the four types being made (namely, Basic, CORS, and Opaque), and just need to figure out how to make an OpaqueRedirect filtered response, since it's handled separately from the others. I will also add more tests to ensure the content of the filtered responses matches the limitations placed by the specification.

Along the way I implemented Cors Check, since it's required for the CORS filtered response. @jdm suggested I handle it in here, since it's such a small step, compared to other parts of Fetch.

Since all the tests currently pass, and I've spent a while adding the Cors Check and other pieces, I figured now would be a good time to start having it reviewed.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9525)

### CODE REVIEW

File: components/net/fetch/methods.rs [hunk 1]
```diff
@@ -861,9 +863,85 @@ fn preflight_fetch(request: Rc<Request>) -> Response {
 /// [CORS check](https://fetch.spec.whatwg.org#concept-cors-check)
 fn cors_check(request: Rc<Request>, response: &Response) -> Result<(), ()> {
     // TODO: Implement CORS check spec
```
Comment (KiChjang) [inline components/net/fetch/methods.rs]: This comment can be removed.

**Comment by nikkisquared** [comment]
I've added tests for Basic and CORS filtered responses to set all headers are are specifically referred to are set for the initial response. Then I checked each of them to test that they were or were not present as specified by Fetch. For the Opaque filtered response I tested that the specified values to reset are properly done.

I could add tests to check the integrity of other headers and values to make sure anything not specified is unaffected, but since there is a huge number of things I could possibly test, I'm not sure where to start.

**Comment by nikkisquared** [comment]
I've added tests for OpaqueRedirect now, and with that I am covering everything the Fetch spec asks for.

I'd like a review to check for any gaps in my coverage, so that I can improve it. When can you review this PR @asajeffrey?

**Comment by asajeffrey** [comment]
OK, comments made, mainly just niggles. The only substantive comment is what the state of play with non-ascii URIs is, what should we do about IRIs?

---

Reviewed 6 of 7 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 868 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nLg0l72zJUjeWrbYO:-K9nLg0l72zJUjeWrbYP:1221008370) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L868)):*
Oops, there's still a debugging println here. It should either be deleted, or turned into a debug log entry.

---

*[components/net/fetch/methods.rs, line 872 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nMQEmWQM8Sx-vNc15:-K9nMQEn_womyPqDS1uU:-1095889910) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L872)):*
`let origin = try!(origin.ok_or(()));` is shorter, if more cryptic.

---

*[components/net/fetch/methods.rs, line 882 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nOKh7eDwNS-dFvMLf:-K9nOKh7eDwNS-dFvMLg:1375153963) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L882)):*
Shouldn't this be part of step 4?

---

*[components/net/fetch/methods.rs, line 890 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nSUQlY0KrZBRs1xA6:-K9nSUQlY0KrZBRs1xA7:423872775) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L890)):*
Should we be ascii serializing the request origin here? The spec doesn't mention which serialization to use. Can the access control origin contain non-ascii characters? Are IRIs supported?

---

*[components/net/fetch/methods.rs, line 908 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nT9fjPGTgM_VdbuR6:-K9nT9fkB4MNnPjUHOJ5:894281923) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L908)):*
Shouldn't this be `if credentials == Some("true")`?

---

*[components/net/fetch/methods.rs, line 922 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nPLsLPmW2b68TIA9Q:-K9nPLsM4xA4AxGgDRzQ:1012755103) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L922)):*
`let host = try!(domain_to_ascii(host.serialize().as_str()))` would bubble the error. Do we really want to discard the error value?

The spec talks about `SecurityError` in step 4, which isn't mentioned here.

---

*[components/net/fetch/methods.rs, line 934 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nQ0jObMzO4UiawyZt:-K9nQ0jObMzO4UiawyZu:-776689942) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L934)):*
Could we avoid re-formatting port twice, with something like

```
            if Some(*port) == default_port {
                format!("{}://{}", scheme, host)
            } else {
                format!("{}://{}:{}", scheme, host, port)
            }
 
```
Note that we shouldn't be including `port` if its the default.

Why do we need to use `Some(*port)` rather than `Some(port)`?

---

*[components/net/fetch/methods.rs, line 936 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nS3L3Y1WOxMVr4Iw1:-K9nS3L3Y1WOxMVr4Iw2:960038216) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L936)):*
The code would be easier to read if this were the first case, and the other cases had comments tying them to steps in the spec.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525#-:-K9nU8VbZn_jW-UJCd6S:-1166888109)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 890 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nSUQlY0KrZBRs1xA6:-K9niQQhJBIGQYuHn8uz:-797099932) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L890)):*
The spec says "If request's origin, serialized and utf-8 encoded, is not origin, return failure.", with "serialized" being a link to https://html.spec.whatwg.org/multipage/browsers.html#ascii-serialisation-of-an-origin.

Since the Access Control Origin header can hold a String[1], it certainly could contain non-ascii characters. However, the spec says under the ascii serialisation step 4, "If the domain to ASCII algorithm returns failure, e.g. ... because it contains invalid characters, then throw a SecurityError exception and abort these steps." I'm under the impression that non-ascii characters are invalid for this header.

I don't know what IRIs are, so I don't know what I can say on them.

[1] http://hyper.rs/hyper/hyper/header/enum.AccessControlAllowOrigin.html

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 908 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nT9fjPGTgM_VdbuR6:-K9njgAaxJjFnaUHN9Yc:-1010248087) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L908)):*
No actually- because of how Hyper simplifies the implementation. The header can only ever have the value `true`, or not exist. So, Hyper doesn't have it store any value[1].

[1] http://hyper.rs/hyper/hyper/header/struct.AccessControlAllowCredentials.html

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 922 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nPLsLPmW2b68TIA9Q:-K9nkiByRfuNXsGpb6DK:-992584625) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L922)):*
@jdm told me that Fetch doesn't have any specific error handling (yet), so there's no way for me to throw a SecurityError in a meaningful way. I'll add a TODO about it though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 936 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nS3L3Y1WOxMVr4Iw1:-K9nlqnSMVoxMICkw1C8:-152283480) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L936)):*
I removed the step comments because the implementation doesn't correlate well to most of the steps, since it's easier to format it all together at the end, rather than one step at a time. I will add
 step comments and we can see how that looks.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 890 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nSUQlY0KrZBRs1xA6:-K9nmpguO5ngD7jHZ1VU:260266894) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L890)):*
I asked in #servo about IRIs and now I understand what they mean. SimonSapin described how the ascii serialisation of an origin works: it converts it to ASCII, instead of rejecting non-ascii characters. Their example was that a call like domain_to_ascii("☃.net") would return the ascii string"xn--n3h.net". So, IRIs should not require any special consideration in the code!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by asajeffrey** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 890 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nSUQlY0KrZBRs1xA6:-K9o7-4Rjvw-SJKUjlMn:-908826163) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L890)):*
Not entirely sure why the spec says "serialized and utf-8 encoded" if it means ASCII serialization, because utf-8 encoding is then a no-op. Is the origin in the header already ASCII serialized?

---

*[components/net/fetch/methods.rs, line 936 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nS3L3Y1WOxMVr4Iw1:-K9o7oS9LkVYM5pMrMwD:1026549904) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L936)):*
OK, lets see what the re-formatted version looks like.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 934 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nQ0jObMzO4UiawyZt:-K9oMByXa3nUvla379bd:-2121927422) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L934)):*
Your suggestion for reformatting it is great!

I don't know why I need to use `Some(*port)`, but it's a compiler error to not: https://gist.github.com/nikkisquared/1507435b994a08122214

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 936 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nS3L3Y1WOxMVr4Iw1:-K9oMUc8eYko5ncDHXCT:-1155537596) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L936)):*
Hey, sorry for the slow reply! I'm working on some build errors and I'll have it up as soon as I can :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by nikkisquared** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 890 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nSUQlY0KrZBRs1xA6:-K9oMjfHI1yNzkNZ7lRp:1769811590) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L890)):*
I'm not sure how to regard the spec either, especially since in Rust it doesn't translate well. utf-8 is what String uses in Rust! I don't know where the origin in the header gets ASCII serialized, but since the spec doesn't ask for me to make sure it is, it likely is handled by Hyper in this situation. Setting the header is not covered under Fetch, regardless.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525)*

**Comment by Ms2ger** [comment]
About utf-8 decoding: the specification assumes you get the value of the `Access-Control-Allow-Origin` header as a sequence of bytes (`[u8]`), and you want to compare it to a string, so you'd have to use `as_bytes()`. However, hyper gives us a string, so comparing them directly is probably the (most) correct thing to do. We might want to confirm that hyper's behaviour is indeed correct, for this header and others, so I'd appreciate it if you'd file a bug for that.

**Comment by asajeffrey** [comment]
The whatwg spec for urls refers to RFC 1034 for domains (https://url.spec.whatwg.org/#biblio-rfc1034) which predates IRIs by quite a bit. A quick google search kicks up (https://tools.ietf.org/html/draft-ruby-url-problem-01) which I would summarize as "this is a mess". The origin header is ASCII, but I think it's allowed to include %-encoded unicode. I think the fetch spec is requiring the origin header to be punycode-encoded rather than %-encoded, though it never actually says so.

I think the thing to do would be to add a couple of small test cases, one which shows a punycode-encoded origin uri passing the CORS check, and one which shows a %-encoded origin URI being rejected (which I think is the spec-compliant behavior even thought it's not what you;'d expect).

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/net/fetch/methods.rs, line 936 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9525#-K9nS3L3Y1WOxMVr4Iw1:-K9rBqXMysYgJfy8pxcV:821537861) ([raw file](https://github.com/servo/servo/blob/4f4d0cac8b8f4514d740209d8be68f6d20c051fe/components/net/fetch/methods.rs#L936)):*
OK, good luck with those build errors.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525#-:-K9r5NCiCCcLSK5H_djE:1262194971)*

**Comment by nikkisquared** [comment]
@asajeffrey @Ms2ger, @jdm suggested I just leave a TODO comment for adding more tests for different URL encodings, and I can come back to that later, or leave it to somebody else. I finally got the Cors Check update to work, thanks to the suggestions on Friday! So I'd like another review on it.

**Comment by asajeffrey** [comment]
Sounds sensible. Might be worth filing an issue about tests for punycode-encoded URLs in Origin.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9525#-:-KA6kQko80FJyRgFVLpA:-1738102153)*

**Comment by asajeffrey** [comment]
Still needs squashing, but apart from that we're good to go!

**Pull request merged**